sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/CoGroup.java

```diff
@@ -107,156 +106,251 @@
+ * identical between input1 and input2.
+ *
+ * <p>This transform also supports outer-join semantics. By default, all input PCollections must
+ * participate fully in the join, providing inner-join semantics. This means that the join will only
+ * produce values for "Bob" if all inputs have values for "Bob;" if even a single input does not
+ * have a value for "Bob," an inner-join will produce no value. However, if you mark that input as
+ * having outer-join participation then the join will contain values for "Bob," as long as at least
+ * one input has a "Bob" value; null values will be added for inputs that have no "Bob" values. To
+ * continue the SQL example:
+ *
+ * <p>SELECT * FROM input1 LEFT OUTER JOIN input2 ON input1.user = input2.user
+ *
+ * <p>Is equivalent to:
+ *
+ * <pre>{@code
+ * PCollection<Row> joined = PCollectionTuple.of("input1", input1, "input2", input2)
+ *   .apply(CoGroup.join("input1", By.fieldNames("user").withOuterJoinParticipation())
+ *                 .join("input2", By.fieldNames("user"))
+ *                 .crossProductJoin();
+ * }</pre>
+ *
+ * <p>SELECT * FROM input1 RIGHT OUTER JOIN input2 ON input1.user = input2.user
+ *
+ * <p>Is equivalent to:
+ *
+ * <pre>{@code
+ * PCollection<Row> joined = PCollectionTuple.of("input1", input1, "input2", input2)
+ *   .apply(CoGroup.join("input1", By.fieldNames("user"))
+ *                 .join("input2", By.fieldNames("user").withOuterJoinParticipation())
+ *                 .crossProductJoin();
+ * }</pre>
+ *
+ * <p>and SELECT * FROM input1 FULL OUTER JOIN input2 ON input1.user = input2.user
+ *
+ * <p>Is equivalent to:
+ *
+ * <pre>{@code
+ * PCollection<Row> joined = PCollectionTuple.of("input1", input1, "input2", input2)
+ *   .apply(CoGroup.join("input1", By.fieldNames("user").withOuterJoinParticipation())
+ *                 .join("input2", By.fieldNames("user").withOuterJoinParticipation())
+ *                 .crossProductJoin();
+ * }</pre>
+ *
+ * <p>While the above examples use two inputs to mimic SQL's left and right join semantics, the
+ * {@link CoGroup} transform supports any number of inputs, and outer-join participation can be
+ * specified on any subset of them.
+ *
+ * <p>Do note that cross-product joins while simpler and easier to program, can cause
  */
 public class CoGroup {
-  /**
-   * Join by the following field names.
-   *
-   * <p>The same field names are used in all input PCollections.
-   */
-  public static Inner byFieldNames(String... fieldNames) {
-    return byFieldAccessDescriptor(FieldAccessDescriptor.withFieldNames(fieldNames));
-  }
+  private static final List NULL_LIST;
 
-  /**
-   * Join by the following field ids.
-   *
-   * <p>The same field ids are used in all input PCollections.
-   */
-  public static Inner byFieldIds(Integer... fieldIds) {
-    return byFieldAccessDescriptor(FieldAccessDescriptor.withFieldIds(fieldIds));
+  static {
+    NULL_LIST = Lists.newArrayList();
+    NULL_LIST.add(null);
```

Comment:
I'm not sure it's going to do safe publishing because list is modified after final value is assigned. What about `NULL_LIST = Arrays.asList((Object) null)`